kordac/processors/HeadingBlockProcessor.py

```diff
@@ -0,0 +1,150 @@
+from markdown.blockprocessors import BlockProcessor
+from markdown.util import etree
+from kordac.utils.HeadingNode import HeadingNode, DynamicHeadingNode
+import re
+
+class HeadingBlockProcessor(BlockProcessor):
+    ''' Searches a Document for markdown headings (e.g. # HeadingTitle) these are then processed into html headings and generates level trails for each.
+    '''
+
+    def __init__(self, ext, *args, **kwargs):
+        '''
+        Args:
+            ext: The parent node of the element tree that children will reside in.
+            args: Arguments handed to the super class.
+            kwargs: Arguments handed to the super class.
+        '''
+        super().__init__(*args, **kwargs)
+        self.processor = 'heading'
+        self.max_levels = 6
+        self.pattern = re.compile(ext.processor_info[self.processor]['pattern'])
+        self.template = ext.jinja_templates[self.processor]
+        self.custom_slugify = ext.custom_slugify
+        self.level_generator = LevelGenerator(self.max_levels)
+
+        self.roots = []
+        self.current_node = None
+        self.update_ext_tree = ext._set_heading_tree
+        self.get_ext_tree = ext.get_heading_tree
+
+    def test(self, parent, block):
+        ''' Tests a block to see if the run method should be applied.
+
+        Args:
+            parent: The parent node of the element tree that children will reside in.
+            block: The block to be tested.
+
+        Returns:
+            True if the block matches the pattern regex of a HeadingBlock.
+        '''
+        return self.pattern.search(block) is not None
+
+    def run(self, parent, blocks):
+        ''' Increments the level trail at the given level. This also
+        zeros any levels after the incremented level.
+
+        Args:
+            level_inc: the level to be incremented.
+
+        Returns:
+            A tuple of levels, where higher levels are first and the lowest
+            level is last.
+        '''
+        block = blocks.pop(0)
+        match = self.pattern.search(block)
+        assert match is not None # If this is true how did we test successfully
```

Comment:
is this line necessary (referencing your own comment) and do we have this check in all the other processors?